usb-c: Add sink VIF policies to device tree, update the GenVIF script

madhurimaparuchuri opened the pull request:
A VIF node is added to device tree to hold VIF specifications and USB sink VIF policies are added to usb-c-connector node as properties.

The GenVIF script is updated to start parsing from VIF node.

The exisiting USB-C Sink sample project is updated to include examples of new changes in device tree.

The XML output is validated against schema
https://compliance.usb.org/cv/VendorInfoFile/Schemas/Current/VendorInfoFile.xsd Changes to XML namespaces is done according to validation

Signed-off-by: Madhurima Paruchuri <mparuchuri@google.com>

nordicjm commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -137,31 +137,152 @@ properties:
+            Swap USB Type-C Current, in which the initial Source describes
+            whether it supports Fast Role Swap to become a new Sink, and if so,
+            what current it requires as a new Sink.
+            Valid options are:
+            0 - FR_Swap not supported
+            1 - Default USB Power
+            2 - 1.5A @ 5V
+            3 - 3A @ 5V
```
enum could optionally be used here

madhurimaparuchuri replied:
I updated the PR with enums. Thank you

tejlmand reviewed (changes requested):
to me this seems like a misuse of devicetree, but I will let @mbolivar-nordic and @galak decide on this.

Waiting for their review comments.

madhurimaparuchuri commented on dts/bindings/usb-c/vif.yaml:
```diff
@@ -0,0 +1,113 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    A VIF node is a root element that describes the Qualifying Product in VIF
+    specification. It contains introductory material followed by an optional
+    product-level field element, zero or more Components, and zero or more
+    Optional Content elements.
+    This is based on VIF XML file schema in:
+      https://compliance.usb.org/cv/VendorInfoFile/Schemas/Current/VendorInfoFile.xsd
+
+compatible: "vif"
```
There are some high level specifications required for VIF, specific to product but applicable to all ports. I see USB-C DT nodes are per port, hence added this node, which is referring to the ports (i.e, usb-c connector nodes).

mbolivar-nordic replied:
@madhurimaparuchuri : 

Without commenting this is appropriate to include in devicetree (I am no longer maintaining the devicetree bindings so that I can focus on other tasks), I will say that choosing "vif" for the compatible is pretty clear namespace pollution. The namespace of compatible property values is shared across all hardware in the world, so a three-letter acronym had better be for something incredibly general purpose, which this definitely is not.

Please choose a better, more descriptive name, and make sure to expand the acronym (I guess it's "vendor info file") in the description field in the binding.

In general, please remember that you are the VIF domain expert here, but this binding lives in a larger world where there's lots of other hardware and people who will read your documentation without your domain knowledge.

mbolivar-nordic replied:
> @galak to review and decide on this

This is the decision maker on whether to put this in DT, since he is overall maintainer for the bindings.

I now only co-maintain the DT infrastructure and API.

tejlmand replied:
> I now only co-maintain the DT infrastructure and API.

I know, but given your experience your feedback serves as very valid input (at least better input than my feedback) on whether this approach looks appropriate for bindings.

alevkoy replied:
I'm inclined to agree that the devicetree should not directly represent the VIF fields for a few reasons:
1. The USB Implementers' Forum revs the VIF format somewhat regularly to keep up with new USB features. Sometimes, existing fields change name or format without changing the underlying meaning. Keeping up with these changes in the devicetree would involve changing a public-facing (sort of) API in the devicetree bindings.
2. Many of the fields will be unknowable to the Zephyr build system, and which ones are knowable depends on the hardware. To produce a valid VIF for, say, compliance testing, a developer would need to add in additional fields from some other source. If I were adding those fields, I'd probably want to use a format other than devicetree.
3. The bindings need to describe the devicetree properties, so they need to be verbose like the spec or refer to the spec or summarize the spec at the risk of vagueness or inaccuracy. I'd rather avoid the issue.
4. The VIF contains some parallel structure to the existing devicetree, e.g. different pieces for different ports. I'd prefer not to replicate that and risk them being inconsistent.

I'd prefer that the nodes for the port and individual port components contain properties for VIF-relevant information that is determined by the hardware, and the VIF generator would parse those properties (probably among other data sources) while constructing the VIF.

madhurimaparuchuri replied:
I avoided adding elements to device tree. I am reading properties that are already there in device tree and generating the VIF, not adding any properties or nodes in this revision. The information that is there in VIF node in earlier revisions is being read form an XML file (manually added) in project folder if it's present (overriding the output VIF file).

madhurimaparuchuri replied:
@tejlmand 
Could you review the new revision and let me know if any changes are required in this PR

gmarull commented on dts/bindings/usb-c/vif.yaml:
```diff
@@ -0,0 +1,113 @@
+            Product. If vif-product-type is not present, then the number of
+            components shall be 0.
+
+    # EXAMPLE:
+    #
+    #
+    # vif {
+    #    address-cells = <1>;
+    #    size-cells = <0>;
+    #    usbcvif1: usbcvif@1 {
+    #       compatible = "vif";
+    #       reg = <1>;
+    #       vendor-name = "Zephyr";
+    #       model-part-number = "b_g474e_dpow1";
+    #       product-revision= "1";
+    #       tid= "0";
+    #       vif-product-type= <0>;
+    #       certification-type= <0>;
+    #       component =<&port1>;
+    #     };
+    # };
```
ditto

madhurimaparuchuri replied:
Included the changes in new revision of this PR

gmarull commented on samples/subsys/usb_c/sink/boards/b_g474e_dpow1.overlay:
```diff
@@ -27,8 +27,28 @@
 			tcpc = <&ucpd1>;
 			vbus = <&vbus1>;
 			power-role = "sink";
+            		num-snk-pdos= <1>;
 			sink-pdos = <PDO_FIXED(5000, 100, 0)>;
+            		fr-swap-reqd-type-c-current-as-initial-source= <0>;
```
missing space before `=`, please revisit formatting on all files.

madhurimaparuchuri replied:
Included the changes in revision of this PR

gmarull commented on samples/subsys/usb_c/sink/boards/b_g474e_dpow1.overlay:
```diff
@@ -27,8 +27,28 @@
+            compatible = "vif";
+            reg = <1>;
+            vendor-name = "Zephyr";
+            model-part-number = "b_g474e_dpow1";
```
there's standard `model` property. Why is this needed?

madhurimaparuchuri replied:
`model-part-number` seemed to be different from model and more specific to USB-C component. Following is it's definition : "To be provided by the Vendor. This field should match what appears in the USB-IF Product Registration form. If the Qualifying Product has a Model Name entry in USB4 DROM, then this field should be identical to that entry. The USB-IF uses this field and /VIF/Product_Revision to generate the product name in the USB Integrator’s List"

madhurimaparuchuri commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -137,31 +137,153 @@ properties:
+            Set to YES if the Qualifying Product may send a PR_Swap request
+            while operating as a Sink.
+
+    # EXAMPLE:
+    #
+    # USB-C connector attached to a STM32 UCPD typec port controller, which has
+    # power delivery support and enables SINK.
+    #
+    # #include <dt-bindings/usb-c/pd.h>
+    #
+    # vbus1: vbus {
+    #     compatible = "zephyr,usb-c-vbus-adc";
+    #     io-channels = <&adc2 8>;
+    #     output-ohms = <49900>;
+    #     full-ohms = <(330000 + 49900)>;
+    # };
+    #
+    # ports {
+    #     #address-cells = <1>;
+    #     #size-cells = <0>;
+    #     port1: usb-c-port@1 {
+    #         compatible = "usb-c-connector";
+    #         reg = <1>;
+    #         tcpc = <&ucpd1>;
+    #         vbus = <&vbus1>;
+    #         power-role = "SINK";
+    #         num-snk-pdos= <2>;
+    #         sink-pdos = <PDO_FIXED(5000, 2000, PDO_FIXED_USB_COMM)
+    #                      PDO_VAR(5000, 12000, 2000)>;
+    #         op-sink-microwatt = <10000000>;
+    #         fr-swap-reqd-type-c-current-as-initial-source= <0>;
+    #         vconn-swap-to-off-supported;
+    #      };
+    # };
```
Should I move this example to the top of this file under description

madhurimaparuchuri replied:
Included the changes in new revision of this PR

madhurimaparuchuri replied:
@gmarull 
Could you review the new revision and let me know if any changes are required in this PR

madhurimaparuchuri commented on dts/bindings/usb-c/vif.yaml:
```diff
@@ -0,0 +1,113 @@
+
+    tid:
+        type: string
+        required: false
```
Included the changes in new revision of this PR

alevkoy commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -137,31 +168,107 @@ properties:
+        type: boolean
+        description: |
+            Set to YES if the Qualifying Product may send a PR_Swap request
+            while operating as a Sink.
```
Some of these properties describe behaviors that will sometimes vary according to runtime circumstances or policy (and thus should probably not be constant port properties).

For instance, whether a port accepts a PR Swap may depend on the battery level or what is plugged into the other ports. FR Swap required current may depend on the power state of the device.

madhurimaparuchuri replied:
I beleive this property is capturing  whether it may send PR_Swap request. I mean it need not to send request consistently, but if it could anytime it's true, kind of.

alevkoy commented on dts/bindings/usb-c/usb-c-connector.yaml:
```diff
@@ -137,31 +168,107 @@ properties:
-  #         op-sink-microwatt = <10000000>;
-  #      };
-  # };
+    epr-supported-as-snk:
```
How did you decide which of these fields to put in the port node vs the VIF node?

madhurimaparuchuri replied:
I was relying on information required to build VIF xml, the properties per port are being added in USB-C connector node, the properties common to all ports are added to VIF node. Following is the VIF XML schema: https://compliance.usb.org/cv/VendorInfoFile/Schemas/Current/VendorInfoFile.xsd , there Component refers to a port. There could be multiple components in a VIF file belonging to a product. Thank you

madhurimaparuchuri commented on samples/subsys/usb_c/sink/vif_b_g474e_dpow1.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version='1.0' encoding='utf-8'?>
+<vif:VIF xmlns:vif="http://usb.org/VendorInfoFile.xsd">
+  <vif:VIF_Specification>3.19</vif:VIF_Specification>
+  <vif:VIF_App>
```
Though the source XML is containing data regarding version and app, it won't add them to output xml. The members mentioned in VIF_SPEC_ELEMENTS_FROM_SOURCE_XML set in constants.py will be picked from source XML.

madhurimaparuchuri commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -129,9 +99,10 @@ def parse_and_add_sink_pdo_to_xml(xml_ele, pdo_value, pdos_info):
-        pdos_info[constants.HIGHER_CAPABILITY_SET] = pdo_value & (1 << 28)
-        pdos_info[constants.FR_SWAP_REQD_TYPE_C_CURRENT_AS_INITIAL_SOURCE] = pdo_value & (3 << 23)
-        add_element_to_xml(xml_ele, constants.SINK_PDO_VOLTAGE, f'{voltage_mv} mV',
+        if pdo_value & (1 << 28):
```
Added the comment in new revision

madhurimaparuchuri replied:
The PD spec version is similar to what USB-C subsystem is using. Table number is changing with PD version, but the higher capability bit is same in different versions. When ever there is change is PD revision in USB-C subsystem, we need to remember this spot and update it. Is it fine adding version and table numbers here then.

alevkoy replied:
> When ever there is change is PD revision in USB-C subsystem, we need to remember this spot and update it.

This is what I'm trying to avoid. If you put the spec version here, someone will be able to find the definitive reference for this code as long as the previous spec versions are available. Without the spec version, they'll have to hope that the table hasn't changed since you wrote this code.

You should only need to update the version listed here if you make changes to the code that depend on spec content that is newer than the current spec version. Otherwise, the current spec version should be an acceptable reference no matter what we do with the USB-C subsystem.

This is still just a nit.

madhurimaparuchuri replied:
Added revision, versions and table no in new revision

alevkoy replied:
PD 2.0 is quite old at this point. Also, I don't think it has APDOs. You should reference PD rev 3.1, v 1.6, table 6-16 here. Battery is 6-18. Variable is 6-17. Augmented is 6-20.

madhurimaparuchuri replied:
I updated the version in new revision. Thank you

alevkoy replied:
Good job double-checking my mistake 2ith 6-20.

madhurimaparuchuri commented on scripts/generate_usb_vif/constants.py:
```diff
@@ -9,81 +9,80 @@
-TEXT = "text"
-ATTRIBUTES = "attributes"
-CHILD = "child"
+# DT node/property names
```
Divided the constant files into multiple in a folder, in-place of classes as it requires initiation.

alevkoy replied:
That seems fine to me, although I'm a novice at Python.

madhurimaparuchuri commented on scripts/generate_usb_vif/constants.py:
```diff
@@ -9,81 +9,80 @@
-FALSE = "false"
 
-SINK_PDOS = "sink-pdos"
+# USB_PD_SUPPORT = "USB_PD_Support"
```
I forgot cleaning up earlier it's removed now

mbolivar-nordic commented on scripts/generate_usb_vif/constants/dt_constnats.py:
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python3
```
Please fix the typo in this file name.

madhurimaparuchuri replied:
Sorry didn't notice it, updated it in new revision. Thank you

alevkoy commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -179,72 +151,69 @@ def parse_and_add_sink_pdo_to_xml(xml_ele, pdo_value, pdos_info):
+                       {other_constants.VALUE: str(len(sink_pdos))})
+
+
+def parse_and_add_component_to_xml(xml_ele, node):
```
Nit: IMO, these long, highly nested functions could use some more blank lines than they have now.

madhurimaparuchuri replied:
Ok, I actually wrote them in one line, when I formatted it in IDE, it broke it that way. Will change it

alevkoy replied:
For example, this function is 20 LOC with 5 levels of indentation. I think it would be easier to follow if there were blank lines before lines 200, 213, and maybe 207.

Similarly, `parse_and_add_sink_pdo_to_xml` is 78 lines long with 4 levels of indentation. I think it would be easier to read with blank lines before lines 101, 107, 123, 140, 160, and 171.

madhurimaparuchuri replied:
Added few lines in new revision

alevkoy commented on scripts/generate_usb_vif/constants/dt_constants.py:
```diff
@@ -0,0 +1,15 @@
+
+DT_VIF_ELEMENTS = {
+    SINK_PDOS: "SnkPdoList",
+    PD_DISABLE: "USB_PD_Support",
```
I find it odd that `USB_PD_Support` has a constant name of `PD_DISABLE`, and it takes its value as the _opposite_ of the `PD_DISABLE` DT property. I would suggest naming the VIF element constant after the VIF elements. Seeing `PD_SUPPORT = not PD_DISABLE` in the code would make more sense intuitively.

Also, it would make more sense to me if these VIF element names went in `vif_element_constants.py. You could import that file from here to use those constant names in `DT_VIF_ELEMENTS`.

(I also find it odd that the DT property is `pd-disable` and not something like `pd-enable` or `pd-support` that would be true when PD is enabled, but that's outside this scope of this PR.)

madhurimaparuchuri replied:
The above dict contains mapping from DT prop names to VIF element name, which code will translate and create XML element. I avoided creating constants in vif_elemnts_constants.py file for them("SnkPdoList", "USB_PD_Support"), as it's used only once and not used by script. 

I will remove the mapping of PD_DISABLE one

madhurimaparuchuri replied:
Removed PD_DISABLE mapping in new revision

madhurimaparuchuri commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -179,72 +151,69 @@ def parse_and_add_sink_pdo_to_xml(xml_ele, pdo_value, pdos_info):
-                           {constants.VALUE: str(pps_max_voltage)})
-        add_element_to_xml(xml_ele, constants.SINK_PDO_OP_CURRENT,
+                           {other_constants.VALUE: str(pps_max_voltage)})
+        add_element_to_xml(xml_ele, vif_element_constants.SINK_PDO_OP_CURRENT,
```
good idea, thank you

madhurimaparuchuri replied:
Updated it in new revision

alevkoy reviewed (dismissed):
LGTM. A few remaining non-blocking comments.

fabiobaltieri commented:
@gmarull @tejlmand hey I think the comments have been addressed, any chance you could give this another round of review?

nordicjm commented on cmake/vif.cmake:
```diff
@@ -5,13 +5,17 @@
         )
 
+if(EXISTS ${APPLICATION_SOURCE_DIR}/vif_${BOARD}.xml)
+    list(APPEND cmd_gen_vif --vif-source-xml ${APPLICATION_SOURCE_DIR}/vif_${BOARD}.xml)
```
This looks like it would get messy with files in the root directory of e.g. a sample - there already is a place for board overlays to go in the `boards` folder. I don't know about vif but is this something you would only ever define at the application level i.e. can this also be defined at a board level just once as opposed to once per board per application, which quickly grows out of control with multiple samples?

madhurimaparuchuri replied:
I got confused/missed thinking about scope of VIF, thank you for pointing out. In the Zephyr sample app "sink", we need different source VIF per board, but there is use-case in ChromeOS where there could be multiple apps using a board(i.e, chip) requires different VIF at application level but not at board. I am evaluating different approaches to be able to have choice to have the VIF in either at application level or board level, will discuss with you offline regarding them. Thank you

madhurimaparuchuri replied:
Hi, I updated the script to look in both application and board directories. Apart from that a user could configure the file location through Kconfig options if they want to store in different folder with different name in application or board source directory. Thank you

fabiobaltieri commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,33 @@
         )
 
+if(CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+    if (EXISTS ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+    elseif (EXISTS ${BOARD_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+        set(vif_source_xml ${BOARD_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+    endif()
```
Is the priority inversion compared to the other case intentional?

fabiobaltieri replied:
Cool, but why?

madhurimaparuchuri replied:
I removed board from config option lookup, as it seems be to not required because the board code is generic for all apps. At present if kconfig option is configured then it will check for that path in application dir. If option is not enabled it will check for vif.xml in board or app directories. The order will enable if vif.xml is defined at both app and board level. The app will have the choice to choose.

fabiobaltieri commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -39,6 +39,17 @@ config BUILD_OUTPUT_VIF
+	string "Path to input VIF file"
+	depends on BUILD_OUTPUT_VIF
+	help
+	  Pass custom path of input VIF XML file residing either in
+	  `APPLICATION_SOURCE_DIR` or `BOARD_DIR` to GenVIF script. Consider
+	  `APPLICATION_SOURCE_DIR` and `BOARD_DIR` as source directory. If this config
```
Is this meant to say that this config is interpreted as a subdiractory of the application path or board path? I don't think it's very clear right now. Also I would avoid referring to `APPLICATION_SOURCE_DIR` or `BOARD_DIR` as that makes it looks like they are `CONFIG_` options, which they are not.

madhurimaparuchuri replied:
Yes, the application directory is considered a source directory for Kconfig option, I removed the variables from description in new revision.

fabiobaltieri commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -39,6 +39,16 @@ config BUILD_OUTPUT_VIF
+	string "Path to input VIF file"
+	depends on BUILD_OUTPUT_VIF
+	help
+	  Pass custom path of input VIF XML file residing in application source
+	  directory to GenVIF script. Consider application source directory as source
+	  directory. If this config option is not configured, the script looks for a
```
I reword with something like "Path of the input VIF XML file relative to the application source directory."

madhurimaparuchuri replied:
I updated the description in new revision

fabiobaltieri commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,31 @@
+else()
+    if (EXISTS ${BOARD_DIR}/${vif_xml})
+        set(vif_source_xml ${BOARD_DIR}/${vif_xml})
+    elseif (EXISTS ${APPLICATION_SOURCE_DIR}/${vif_xml})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${vif_xml})
```
Back to the original comment from Jamie, should be really be `<application>/vif.xml`? That is: does it make sense for a vif.xml to be shared with multiple boards? It may make more sense if it would lookup in `<application>/boards/<board>_vif.xml` or something like that.

Look for how other samples use the `boards` directory.

madhurimaparuchuri commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,31 @@
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+    endif()
+else()
+    if (EXISTS ${BOARD_DIR}/${vif_xml})
+        set(vif_source_xml ${BOARD_DIR}/${vif_xml})
+    elseif (EXISTS ${APPLICATION_SOURCE_DIR}/${vif_xml})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${vif_xml})
```
If user has vif file in app directory and board directory with same default filename i.e, "vif.xml" and they need board one to be picked, this order of lookup will pick it. If they need application one they can configure it through Kconfig option. But having application dir looked up first will not support that case.

madhurimaparuchuri commented on subsys/usb/usb_c/Kconfig:
```diff
@@ -39,6 +39,13 @@ config BUILD_OUTPUT_VIF
 	help
 	  Generate XML file containing VIF policies during project build.
 
+config GENVIF_INPUT_VIF_XML_PATH
+	string "Path to input VIF file"
+	depends on BUILD_OUTPUT_VIF
+	help
+	  Path of the GenVIF input VIF XML file relative to the application source
+	  directory.
+
```
GENVIF_INPUT_VIF_XML_PATH holds path to a single xml file like "subboard/src/sink_vif.xml". At present it doesn't support multiple files. The single file path configured in prj.conf will be looked up in application directory.

This config options comes handy when project structure of users is complex. I mean storing vif source file in subdirectories with different names.

tejlmand replied:
> GENVIF_INPUT_VIF_XML_PATH holds path to a single xml file like "subboard/src/sink_vif.xml". At present it doesn't support multiple files. The single file path configured in prj.conf will be looked up in application directory.

not sure the question was fully understood.
So let me try to rephrase.

Use case 1:
- prj.conf: GENVIF_INPUT_VIF_XML_PATH=vif.xml
- prj_alt1.conf: GENVIF_INPUT_VIF_XML_PATH=vif_alt1.xml
- prj_alt2.conf: GENVIF_INPUT_VIF_XML_PATH=vif_alt2.xml
That is, users build with a given conf file which then points to a given vif xml, for example:
```
west build .... (uses prj.conf, and thereby vif.xml)
west build -DCONF_FILE=prj_alt2.conf .... (uses prj_alt2.conf, and thereby vif_alt2.xml)
```

Use case 2:
- prj.conf: GENVIF_INPUT_VIF_XML_PATH=vif.xml
- vif_alt1.xml
- vif_alt2.xml
Users are expected to build with prj.conf, but adjust vif xml using GENVIF_INPUT_VIF_XML_PATH, for example:
```
west build .... (uses prj.conf, and thereby vif.xml)
west build -DGENVIF_INPUT_VIF_XML_PATH=vif_alt2.xml .... (uses prj.conf, but with vif_alt2.xml instead)
```

Use case 3:
- prj.conf: GENVIF_INPUT_VIF_XML_PATH=vif.xml
- boards/foo.conf: GENVIF_INPUT_VIF_XML_PATH=boards/vif_foo.xml
Users are expected to build build for a board, and the board config fragment will point to alternative vif xml when required.
This means:
```
west build -DBOARD=<someboard> .... (uses prj.conf, and thereby vif.xml)
west build -DBOARD=foo .... (uses boards/foo.conf with prj.conf, and thereby boards/vif_foo.xml is used)
```

i'm not familiar with the vif and the intentions / under which circumstances the vif.xml would require changes for a given board.
How static is a vif.xml expected to be for a given application / application configuration ?
Is vif.xml expected to mostly be coupled to an application or to a board configuration.

Could you provide a couple of examples on when, how, why a vif.xml would change ?

Without such background knowledge it becomes very hard to review this PR and to provide best guidance on how to integrate this feature into the build system.

tejlmand replied:
Thanks for following up:
> For use case 1
> ...
> For use case 2, 3

I was not asking how it could be done, I was trying to understand what are the most likely / expected use-case that users will deal with.
You are free to pick one of the sketched use-cases, but I hoped you could describe the use-case intended, cause having an idea about most common use-case ( != all possible use-cases) then that give a better foundation for deciding correct approach.

> And I see in Zephyr based EC of ChromeOS, one board will be used by multiple projects, and USB C sub systems differs at project level.

Now this looks like what i'm asking for :+1: 
When you refer to `project1`, then i'm assuming that's a dedicated `prj.conf`, perhaps pointed to by `CONF_FILE=<file>` or using `APPLICATION_CONFIG_DIR=<dir>` which redirects the whole configuration setup to a dedicated folder, correct ?
So this is very similar to use case 1 in the above, would you agree ?

>  VIF is more specific to application or USBC subsystem it contains.

From this sentence and other comments, I have a slight feeling that it will be very hard to create a board default `vif.xml` that is suitable in most applications.
As the `vif.xml` is xml, and not like dts overlay files and config fragments, then it cannot as easily be extended.
I therefore start to wonder if a board default `vif.xml` is really that useful ?

Maybe application w/ board specific adjustments is all that really is needed, as was proposed here:
https://github.com/zephyrproject-rtos/zephyr/pull/52759#discussion_r1085022956
https://github.com/zephyrproject-rtos/zephyr/pull/52759#discussion_r1112951561
> > It may make more sense if it would lookup in <application>/boards/<board>_vif.xml or something like that.

but you actually did a different implementation than what was proposed.
You decided to use `BOARD_DIR` which is located in `<zephyr_base>/boards/<arch>/<board>`, which is not the same as `<application>/boards` that was proposed.

> it's user choice how they want to organise projects

correct, but having consistency in how Zephyr does things will make it easier to understand and provide a good starting point.
Hence the preference to keep an application specific file taking precedence over a board default from `BOARD_DIR`.
A board default in `<application>/boards` should take precendence over application default, and that is what was initially proposed.

madhurimaparuchuri replied:
Thank you for the revert

`When you refer to project1, then i'm assuming that's a dedicated prj.conf, perhaps pointed to by CONF_FILE=<file> or using APPLICATION_CONFIG_DIR=<dir> which redirects the whole configuration setup to a dedicated folder, correct ?
So this is very similar to use case 1 in the above, would you agree ?`
yes, I agree

`As the vif.xml is xml, and not like dts overlay files and config fragments, then it cannot as easily be extended.
I therefore start to wonder if a board default vif.xml is really that useful ?`
If USB-C system is dedicated to board and at application level nothing related to UBS-C changes, then in that case it will be helpful

`but you actually did a different implementation than what was proposed.
You decided to use BOARD_DIR which is located in <zephyr_base>/boards/<arch>/<board>, which is not the same as <application>/boards that was proposed.`
As project structure seemed to be changing per user, I skipped looking for app/boards/board_vif.xml in cmake script as Kconfig option covers that case.

madhurimaparuchuri commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,31 @@
         )
 
+if(CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+    if (EXISTS ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
```
Added this in new revision

tejlmand replied:
still uses `APPLICATION_SOURCE_DIR`

madhurimaparuchuri replied:
Sorry my bad I missed it, will correct in new revision

madhurimaparuchuri commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -255,9 +253,33 @@ def parse_args():
+        sys.exit(f"devicetree error: {err}")
+
+    xml_root = get_root()
+    source_xml_root = get_source_xml_root(args.vif_source_xml)
```
Added this in new revision

madhurimaparuchuri commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,31 @@
 
+if(CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+    if (EXISTS ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
```
Made it a mandatory parameter, if nothing gets passed the build will break

tejlmand replied:
Seems you missed this part:
> The CMake code can then be simplified.

Code stilkl suffers from the fact that if user makes a typo, like `oof.xml` instead of the correct `foo.xml` (which is the file existing on disk), then `vif_source_xml` is unset and no file argument is passed to python script.

Hence no build failure and user may not notice a mistake was made.

madhurimaparuchuri replied:
Ok, if this `vif_source_xml` parameter is included in command, it will never be empty string as we are appending it to the command only when we are sure file exists. I made this parameter mandatory while reading the command by genVIF script. If it's not included in the command. it will fail the script execution, which stops the build.

madhurimaparuchuri commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -179,72 +167,82 @@ def parse_and_add_sink_pdo_to_xml(xml_ele, pdo_value, pdos_info):
+                       get_value_attribute(snk_max_power))
+    add_element_to_xml(xml_ele, vif_element_constants.HIGHER_CAPABILITY_SET,
+                       None, get_value_attribute(get_xml_bool_value(
+            vif_element_constants.HIGHER_CAPABILITY_SET in pdo_info)))
```
Updated it in new revision

alevkoy commented on scripts/generate_usb_vif/generate_vif.py:
```diff
@@ -255,9 +252,34 @@ def parse_args():
+
+    xml_root = get_root()
+    source_xml_root = None
+    if args.vif_source_xml:
```
If this arg is `required`, do you need to check this here?

madhurimaparuchuri replied:
Yes it's not required, removing in new revision

madhurimaparuchuri committed:
USB-C: genVIF: Cleanup and add support to pick static data from input

Removed few VIF properties which are being hardcoded
Updated the script to parse source VIF XML and add information to
the output
Added optional Kconfig option to configure custom source VIF XML path
Cleaned up the code

Signed-off-by: Madhurima Paruchuri <mparuchuri@google.com>

madhurimaparuchuri commented:
I was trying to push upstream rebase first and commit(my changes) next. As part of it I made soft reset of last commit, when I pushed it, this pull request is closed.  I am trying to reopen it. Meanwhile this commit has latest changes which I was trying to push here to resolve comments : https://github.com/zephyrproject-rtos/zephyr/compare/main...madhurimaparuchuri:zephyr:gvif

madhurimaparuchuri commented:
The related commit is added back to this PR, but the commit looks like a whole new change now. Sorry for messing the PR, it's unintentional.

tejlmand commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,37 @@
+        --vif-out ${vif_out}
         )
 
+if(CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+    if(IS_ABSOLUTE ${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+        if (EXISTS ${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+            set(vif_source_xml ${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
```
I made this comment a month ago which is still not addressed: https://github.com/zephyrproject-rtos/zephyr/pull/52759#discussion_r1144390785
> > The CMake code can then be simplified.
>
> Code stilkl suffers from the fact that if user makes a typo, like oof.xml instead of the correct foo.xml (which is the file existing on disk), then vif_source_xml is unset and no file argument is passed to python script.
>
> Hence no build failure and user may not notice a mistake was made.

Please fix.

madhurimaparuchuri replied:
The build breaks if there is a typo in the file path configured at present

As discussed offline, I added the error messages in CMAKE to say the build failed due to file not found in new revision

tejlmand replied:
I see you added an extra check below, accepted.

tejlmand reviewed (changes requested):
please fix ensure old comments are addressed.

tejlmand commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,43 @@
+        set(vif_source_xml ${APPLICATION_CONFIG_DIR}/${CONFIG_GENVIF_INPUT_VIF_XML_PATH})
+    endif()
+else()
+    if (EXISTS ${APPLICATION_SOURCE_DIR}/boards/${BOARD}_${vif_xml})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/boards/${BOARD}_${vif_xml})
+    elseif (EXISTS ${APPLICATION_SOURCE_DIR}/${vif_xml})
+        set(vif_source_xml ${APPLICATION_SOURCE_DIR}/${vif_xml})
```
those should also be updated to use `APPLICATION_CONFIG_DIR` instead of `APPLICATION_CONFIG_DIR` to be consistent with other configuration files.

madhurimaparuchuri replied:
Included the changes in new revision

tejlmand commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,43 @@
+    if(CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+        message(FATAL_ERROR "Incorrect VIF source XML file path")
+    else()
+        message(FATAL_ERROR "No VIF source XML file found")
```
Please some extra information on how to fix this situation.
Could be something along:
```
To fix, create 'vif.xml' in the application directory or specify a vif.xml using 'CONFIG_GENVIF_INPUT_VIF_XML_PATH'
```

madhurimaparuchuri replied:
Included the changes in new revision, Thank you

tejlmand reviewed (changes requested):
looks better, some minor nits.

Seems only a single APPLICATION_SOURCE_DIR was updated to APPLICATION_CONFIG_DIR wrt. vif xml file.
That should of course be all occurrences in order to be consistent with general conf file behavior.

tejlmand commented on cmake/vif.cmake:
```diff
@@ -4,14 +4,44 @@
+    if (CONFIG_GENVIF_INPUT_VIF_XML_PATH)
+        message(FATAL_ERROR "Incorrect VIF source XML file path. To fix specify correct XML file path at 'CONFIG_GENVIF_INPUT_VIF_XML_PATH'.")
+    else ()
+        message(FATAL_ERROR "No VIF source XML file found. To fix, create '${board_vif_xml}' in boards directory of application directory, or create '${vif_xml}' file in application directory or board directory, or supply a custom XML VIF path using 'CONFIG_GENVIF_INPUT_VIF_XML_PATH'.")
```
Please line wrap.

CMake allows multiple strings with `message()` and will concatenate those when printing.
```suggestion
        message(FATAL_ERROR "No VIF source XML file found. "
                "To fix, create '${board_vif_xml}' in boards directory of application directory, "
                "or create '${vif_xml}' file in application directory or board directory, "
                "or supply a custom XML VIF path using 'CONFIG_GENVIF_INPUT_VIF_XML_PATH'."
        )
```

fabiobaltieri commented:
@gmarull could you revisit this?
